// @sanity/visual-editing/dist/_chunks-es/VisualEditingComponent.js
import { c } from "react/compiler-runtime";
import { useRouter } from "next/router.js";
import { useState, useEffectEvent, useEffect } from "react";
import { enableVisualEditing } from "./enableVisualEditing.js";
function VisualEditingComponent(props) {
  const $ = c(29), {
    components,
    keepStegaOnCopy,
    refresh: refreshProp,
    zIndex,
    onPerspectiveChange,
    onSuspiciousStega: onSuspiciousStegaProp
  } = props, router = useRouter(), [navigate, setNavigate] = useState();
  let t0;
  $[0] !== refreshProp || $[1] !== router ? (t0 = (payload) => {
    if (refreshProp)
      return refreshProp(payload);
    const routerRefresh = async function() {
      await router.replace(router.asPath, void 0, {
        scroll: !1,
        shallow: !1
      });
    }, skipRefresh = _temp, mutationRefresh = () => (console.debug("No loaders in live mode detected, reloading server props"), routerRefresh());
    switch (payload.source) {
      case "manual":
        return routerRefresh();
      case "mutation":
        return payload.livePreviewEnabled ? skipRefresh() : mutationRefresh();
      default:
        throw new Error("Unknown refresh source", {
          cause: payload
        });
    }
  }, $[0] = refreshProp, $[1] = router, $[2] = t0) : t0 = $[2];
  const refresh = useEffectEvent(t0);
  let t1;
  $[3] !== router ? (t1 = (event) => {
    switch (event.type) {
      case "push":
        return router.push(event.url);
      case "pop":
        return router.back();
      case "replace":
        return router.replace(event.url);
      default:
        throw new Error(`Unknown event type: ${event.type}`);
    }
  }, $[3] = router, $[4] = t1) : t1 = $[4];
  const update = useEffectEvent(t1);
  let t2;
  $[5] !== onSuspiciousStegaProp ? (t2 = (reports) => onSuspiciousStegaProp?.(reports), $[5] = onSuspiciousStegaProp, $[6] = t2) : t2 = $[6];
  const onSuspiciousStega = useEffectEvent(t2), hasSuspiciousStegaCallback = typeof onSuspiciousStegaProp == "function";
  let t3;
  $[7] !== components || $[8] !== hasSuspiciousStegaCallback || $[9] !== keepStegaOnCopy || $[10] !== onPerspectiveChange || $[11] !== onSuspiciousStega || $[12] !== refresh || $[13] !== update || $[14] !== zIndex ? (t3 = () => {
    const disable = enableVisualEditing({
      components,
      keepStegaOnCopy,
      zIndex,
      refresh,
      onPerspectiveChange,
      onSuspiciousStega: hasSuspiciousStegaCallback ? onSuspiciousStega : void 0,
      history: {
        subscribe: (_navigate) => (setNavigate(() => _navigate), () => setNavigate(void 0)),
        update
      }
    });
    return () => disable();
  }, $[7] = components, $[8] = hasSuspiciousStegaCallback, $[9] = keepStegaOnCopy, $[10] = onPerspectiveChange, $[11] = onSuspiciousStega, $[12] = refresh, $[13] = update, $[14] = zIndex, $[15] = t3) : t3 = $[15];
  let t4;
  $[16] !== components || $[17] !== hasSuspiciousStegaCallback || $[18] !== keepStegaOnCopy || $[19] !== onPerspectiveChange || $[20] !== zIndex ? (t4 = [components, keepStegaOnCopy, zIndex, onPerspectiveChange, hasSuspiciousStegaCallback], $[16] = components, $[17] = hasSuspiciousStegaCallback, $[18] = keepStegaOnCopy, $[19] = onPerspectiveChange, $[20] = zIndex, $[21] = t4) : t4 = $[21], useEffect(t3, t4);
  const {
    asPath,
    basePath,
    locale,
    isReady
  } = useRouter();
  let t5, t6;
  return $[22] !== asPath || $[23] !== basePath || $[24] !== isReady || $[25] !== locale || $[26] !== navigate ? (t5 = () => {
    if (navigate && isReady) {
      const url = basePath || locale ? `${basePath}${locale ? `/${locale}` : ""}${asPath === "/" ? "" : asPath}` : asPath;
      navigate({
        type: "push",
        url
      });
    }
  }, t6 = [asPath, basePath, isReady, locale, navigate], $[22] = asPath, $[23] = basePath, $[24] = isReady, $[25] = locale, $[26] = navigate, $[27] = t5, $[28] = t6) : (t5 = $[27], t6 = $[28]), useEffect(t5, t6), null;
}
function _temp() {
  return console.debug("Live preview is setup, mutation is skipped assuming its handled by the live preview"), !1;
}
export {
  VisualEditingComponent as default
};
