// @sanity/visual-editing/dist/_chunks-es/VisualEditingComponent2.js
import { c } from "react/compiler-runtime";
import { useRef, useState, useEffect, useEffectEvent, startTransition } from "react";
import { useNavigate, useRevalidator, useLocation } from "react-router";
import { enableVisualEditing } from "./enableVisualEditing.js";
function VisualEditingComponent(props) {
  const $ = c(33), {
    components,
    keepStegaOnCopy,
    refresh,
    zIndex,
    onPerspectiveChange,
    onSuspiciousStega: onSuspiciousStegaProp
  } = props, navigateRemix = useNavigate(), navigateRemixRef = useRef(navigateRemix), [navigate, setNavigate] = useState(), revalidator = useRevalidator(), [revalidatorPromise, setRevalidatorPromise] = useState(null), [revalidatorLoading, setRevalidatorLoading] = useState(!1);
  let t0, t1;
  $[0] !== navigateRemix ? (t0 = () => {
    navigateRemixRef.current = navigateRemix;
  }, t1 = [navigateRemix], $[0] = navigateRemix, $[1] = t0, $[2] = t1) : (t0 = $[1], t1 = $[2]), useEffect(t0, t1);
  let t2, t3;
  $[3] !== revalidator.state || $[4] !== revalidatorLoading || $[5] !== revalidatorPromise ? (t2 = () => {
    revalidatorPromise && revalidator.state === "loading" ? startTransition(() => setRevalidatorLoading(!0)) : revalidatorPromise && revalidatorLoading && revalidator.state === "idle" && (revalidatorPromise(), startTransition(() => {
      setRevalidatorPromise(null), setRevalidatorLoading(!1);
    }));
  }, t3 = [revalidatorLoading, revalidator.state, revalidatorPromise], $[3] = revalidator.state, $[4] = revalidatorLoading, $[5] = revalidatorPromise, $[6] = t2, $[7] = t3) : (t2 = $[6], t3 = $[7]), useEffect(t2, t3);
  let t4;
  $[8] !== onSuspiciousStegaProp ? (t4 = (reports) => onSuspiciousStegaProp?.(reports), $[8] = onSuspiciousStegaProp, $[9] = t4) : t4 = $[9];
  const onSuspiciousStega = useEffectEvent(t4), hasSuspiciousStegaCallback = typeof onSuspiciousStegaProp == "function";
  let t5;
  $[10] !== components || $[11] !== hasSuspiciousStegaCallback || $[12] !== keepStegaOnCopy || $[13] !== onPerspectiveChange || $[14] !== onSuspiciousStega || $[15] !== refresh || $[16] !== revalidator || $[17] !== zIndex ? (t5 = () => {
    const disable = enableVisualEditing({
      components,
      keepStegaOnCopy,
      zIndex,
      onPerspectiveChange,
      onSuspiciousStega: hasSuspiciousStegaCallback ? onSuspiciousStega : void 0,
      refresh: (payload) => {
        const refreshDefault = function() {
          return payload.source === "mutation" && payload.livePreviewEnabled ? !1 : new Promise((resolve) => {
            revalidator.revalidate(), setRevalidatorPromise(() => resolve);
          });
        };
        return refresh ? refresh(payload, refreshDefault) : refreshDefault();
      },
      history: {
        subscribe: (_navigate) => (setNavigate(() => _navigate), () => setNavigate(void 0)),
        update: (update) => {
          update.type === "push" || update.type === "replace" ? navigateRemixRef.current(update.url, {
            replace: update.type === "replace"
          }) : update.type === "pop" && navigateRemixRef.current(-1);
        }
      }
    });
    return () => disable();
  }, $[10] = components, $[11] = hasSuspiciousStegaCallback, $[12] = keepStegaOnCopy, $[13] = onPerspectiveChange, $[14] = onSuspiciousStega, $[15] = refresh, $[16] = revalidator, $[17] = zIndex, $[18] = t5) : t5 = $[18];
  let t6;
  $[19] !== components || $[20] !== hasSuspiciousStegaCallback || $[21] !== keepStegaOnCopy || $[22] !== onPerspectiveChange || $[23] !== refresh || $[24] !== revalidator || $[25] !== zIndex ? (t6 = [components, keepStegaOnCopy, refresh, revalidator, zIndex, onPerspectiveChange, hasSuspiciousStegaCallback], $[19] = components, $[20] = hasSuspiciousStegaCallback, $[21] = keepStegaOnCopy, $[22] = onPerspectiveChange, $[23] = refresh, $[24] = revalidator, $[25] = zIndex, $[26] = t6) : t6 = $[26], useEffect(t5, t6);
  const location = useLocation();
  let t7, t8;
  return $[27] !== location.hash || $[28] !== location.pathname || $[29] !== location.search || $[30] !== navigate ? (t7 = () => {
    navigate && navigate({
      type: "push",
      url: `${location.pathname}${location.search}${location.hash}`
    });
  }, t8 = [location.hash, location.pathname, location.search, navigate], $[27] = location.hash, $[28] = location.pathname, $[29] = location.search, $[30] = navigate, $[31] = t7, $[32] = t8) : (t7 = $[31], t8 = $[32]), useEffect(t7, t8), null;
}
export {
  VisualEditingComponent as default
};
